### DISCUSSION TITLE
Query database from my Go source code

### DISCUSSION THREAD

**Discussion by jerome-diver** [rpm-software-management/rpm#3308] [General]
Hi @pmatilai (or any other maintainer that want to talk about that), i want to create a general kind of package viewer (not to install anything, but just show) than can embed a lot of other packages managers (not only the one for ditro, but also for snap and flatpak containers, and language specific packages/modules managers). And iḿ doing this in Go.
## Query from my own code
I don´t want to do something dirty like to exec some shell command from my own code. Instead, i read files generated by managers and/or databases (for dnf, it is easy).

So there is 2 (or maybe more) options:
1. read the rpmdb.sqlite sqlite3 database (or the Packages.db from openSuse from ndb database)
1. use a golang api that can access/query rpm safely.

## I want to show
- the package name
- his status (installed or present in a repo)
- his version (full version)
- date of installation
- date of last update
- packages dependencies used by the package applicatrion(s)
- packages dependencies that <"can be used" or "are using"> by an other one     

## Question
1. From the first option: is there a kind of UML database diagram that can show and explain what i can read or should i do it myself ?
2. From the second option: do you know an go API that can query the way you probably like the most (me too) the rpm things ?
3. An other one idea is very welcome (but sure not to exec shell code from source code), do you have some better options around ?

**Comment by pmatilai** [comment]
Upstream rpm provides a C API and Python bindings, that's how you access the rpmdb and anything else in rpm. 3rd party bindings exist for at least Perl and Rust. No idea about Go, but you could always roll your own.

https://github.com/rpm-software-management/rpm/tree/master/python/examples has some examples for querying, it's of course Python but gives you an idea what the rpmdb query API looks like.

**Reply by jerome-diver** [reply]
Ok, i just have a quick look at the Python code that use the C API parser. The Python code is really nice and easily readable> The C code request to me a bit more time than 5 minutes (probably one or two days) to get the point about how to interface the API, so it is good.
I didn´t find any API parser on Go, but maybe i can just do the same the Go way.
I will tell you what was my choice (if i decide to build a Go rpm API in the same way, i think you would like to read it...).

**Reply by jerome-diver** [reply]
Hi @pmatilai, from a golang code, i can use a C code. But i'm trying to understand how to query the database from the rpmlib C API.
For example, i read this documentation: [ftp.rpm looks like official but too old and i failed to read for whish version it is valuable](http://ftp.rpm.org/max-rpm/s1-rpm-rpmlib-example-code.html)
But the code is no more good for rpm-4.19.1.1
Do you know and want to share an official and maitained documentation that can be used to learn librpm C code to use ? Or do you know someone that know where i can read something and learn about this ? It is now very loing time and difficult to just learn to use C API librpm to me.

**Reply by pmatilai** [reply]
@jerome-diver, all this harping about the C API docs in multiple totally unrelated places like the evolution of rpm format blog and whatnot, it doesn't fill me with a desire to help you. It does exact the opposite. You may want to think about that a bit.

Max-RPM is well over 20 years old now, most of it isn't very relevant today. The API documentation linked from https://rpm.org/documentation.html is as maintained and as official as it's going to get because it's derived straight from the rpm source. Tutorial, we have not.

**Reply by jerome-diver** [reply]
## "all this rehashing... in totally unrelated places... it doesn't fill me with a desire to help you. It does exact the opposite. You may want to think about that a bit."

I think we all have the ability to think a little but also to have sympathetic wills or not to imagine hard things. I have tried many times to get by with this API on my own, I have not lacked efforts for this for months. Maybe I am too stupid to, in your eyes, have the right to ask, to dare to try to be visible. I do not see where is wrong in this approach but it would seem that I am at your mercy and that you are the only judge here of what is good or bad. You judge me harshly and lecture me now... ok, what can I do? I don't ask you for a favor, do you? Your answer makes it seem like I should beg for a favor... wow, is that really how it is for you?


## "The API documentation linked from https://rpm.org/documentation.html is as maintained and as official as it's going to get because it's derived straight from the rpm source. Tutorial, we have not."

I'm just trying to find a MANUAL ( i can not imagine a tutorial from you, sure... come on... i didn't ask so much) that explains HOW TO USE this API (not just a quick summary of each function as a poor comment, but how to use them together to get something, i can not imagine you do not get the point).

**Please don't give me this show anymore, it's not necessary, you haven't wanted to answer for months, ok... but maybe someone else will be nicer, why not? I'm often nice to others... it makes me happy and I don't grant it as a favor.** I feel a kind of shame on you now. I did nothing wrong. Calm down. I'm not authistic too, so sorry if i can not understand your reaction.


## "Max-RPM is well over 20 years old now, most of it isn't very relevant today."

Yes, sure... **why is it still online and no indication about wish version to use with ?** I do not understand (yeah... you told me indirectly that i am stupid... but you still no explain why is it still there if it is not relevant). What is your problem with me ? Mine is just to find a relevant manual documentation and to point how not relevant one are a nightmare in the process to learn something. You seems to be intelligent and provide lessons there... so you can understand, yes ? What's wrong with you. 

## "it doesn't feel me with a desire..."

OMG... i'm very sorry, you are so high in the sky... i can not believe it dude. Why do you react like that ? Do you feel better after this unfriendly reaction ? If you do not want to help or to do something with that, you absolutely don't need to find an excuse or a ghost or to be rude with me, it is not welcome, i realy don't appreciate that.

I will go ahead to try to find an help, you are not God or my Boss, and please, don't invent something dirty around, it will not help no one there. But sure, people would be afraid to ask something to you after that.

Why not just answer: "i do a lot there to maintain things, but i feel lazy with documentation because as long as I understand myself alone so it feels me like perfect as i am..."
And you know what ? I will not be rude for that kind of thinking, i'm not an asshole. Nothing wrong to be lazy or to imagine that all that you wrote is perfect. What is not normal is to try to find an excuse and be rude with me because i'm just searching an help since month. Come on baby... really ? It looks like a fiction. It's better to laugh about this drama, it's not worth anything humanly good.

**Reply by pmatilai** [reply]
It seems you didn't get my message at all. Had it been just [this simple question](https://github.com/rpm-software-management/rpm/discussions/3308#discussioncomment-11141929) I would've merrily answered it. I *still* answered it. But you have been making loud demands about this on several unrelated topics. Sorry but that's not just how this works. Annoying the people who might be able to help you tends to be counter-productive is what I'm saying.

**Reply by jerome-diver** [reply]
ok, sorry if i get it like it was too high. I'm not english native and maybe your answer was not so haughty as i feel it was.
So there is nothing to explain how we are supposed to use this API and nothing will be done in this way. Ok, thank you and have a good rest.
